### PR TITLE
[VM/VMSS] Fix idempotency for VM/VMSS create

### LINKED ISSUE #3563: OS Disk name for vm create is not stable -> command is not idempotent
From the vm_create function:

    ...
    os_disk_name = os_disk_name or 'osdisk_{}'.format(random_string(10))
    ....

This means that if I invoke the command twice with the same parameters (with os disk name omitted), I will (with a very high probability :)) end up with two disks, one of which is orphaned. This goes against the goal of having idempotent commands.

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes #3563.  Uses a hash algorithm to generate a deterministic hash. For VM and VMSS create, we hash the ID as that is sufficient, but you could seed it with additional strings if needed.

Added tests for VM and VMSS create to verify idempotency, as well as unit tests of the "hash_string" utility method.

I've checked the other template creates and they did not use the "random_string" method, so they are fine.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [X] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [X] Each command and parameter has a meaningful description.
- [N/A] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by tjprescott** [comment]
ping @troydai @johanste @yugangw-msft

File: src/azure-cli-core/tests/test_util.py [hunk 1]
```diff
@@ -141,6 +142,44 @@ def test_shell_safe_json_parse(self):
...
+                # come before...
+                test_lists2 = test_lists[test_lists.index(item1):]
+                for item2 in test_lists2:
+                    result2 = hash_string(item2, length, force_lower, digits_only)
+                    if item1 == item2:
+                        self.assertEqual(result1, result2)
+                        equal_count += 1
+                    else:
+                        self.assertNotEqual(result1, result2)
+            self.assertEqual(equal_count, len(test_lists))
```
Comment (yugangw-msft) [inline src/azure-cli-core/tests/test_util.py]: With the algorithm, collisions will happen. But when the test failed for that, I don't think we can do anything.  Also on failures the test doesn't dump out the random content, so it is hard to follow up, other than just resetting the test.
So I don't think we need this test scenario based on random strings, rather just add a few real world scenario testing using disk, nsg names, etc.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py [hunk 2]
```diff
@@ -1546,7 +1550,7 @@ def create_vm(vm_name, resource_group_name, image=None, size='Standard_DS1_v2',
     vm_dependencies = []
     if storage_account_type == 'new':
         storage_account = storage_account or 'vhdstorage{}'.format(
-            random_string(14, force_lower=True))
+            hash_string(vm_id, length=14, force_lower=True))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: Note, for storage account, we don't need this logic, because the subsequent `vm create` will find and reuse the account. If this is for code consistency, then it is fine.

File: src/azure-cli-core/azure/cli/core/util.py [hunk 3]
```diff
@@ -214,3 +214,22 @@ def random_string(length=16, force_lower=False, digits_only=False):
...
+        try:
+            m.update(c)
+        except TypeError:
+            m.update(c.encode())
+    digest = m.hexdigest()
+    digest = digest.lower() if force_lower else digest
+    digest = str(int(digest, 16)) if digits_only else digest
+    while len(digest) < length:
+        digest = digest + digest
+    return digest[:length]
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/util.py]: I agree--if we see it failing in CI builds frequently, I'll adjust it.

File: src/azure-cli-core/tests/test_util.py (hunk 1 above)
Comment (tjprescott) [inline src/azure-cli-core/tests/test_util.py]: The reason I use random strings is to see how often collisions occur to see if it really is a significant issue. If I used carefully chosen values, it would appear that there aren't any collisions.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py (hunk 2 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/custom.py]: That does raise the issue in #3589.  Finding and reusing the storage account 1) isn't guaranteed (maybe it finds a different storage account) and 2) is arguable whether it is idempotent since running the same command twice results in different ARM templates.

Here yes I basically did it for consistency.

File: src/azure-cli-core/azure/cli/core/util.py [hunk 4]
```diff
@@ -214,3 +214,22 @@ def random_string(length=16, force_lower=False, digits_only=False):
     if not digits_only:
         choice_set += ascii_lowercase if force_lower else ascii_letters
     return ''.join([choice(choice_set) for _ in range(length)])
+
+
+def hash_string(components, length=16, force_lower=False, digits_only=False):
```
Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/util.py]: Done.

File: src/azure-cli-core/azure/cli/core/util.py [hunk 5]
```diff
@@ -214,3 +214,18 @@ def random_string(length=16, force_lower=False, digits_only=False):
     if not digits_only:
         choice_set += ascii_lowercase if force_lower else ascii_letters
     return ''.join([choice(choice_set) for _ in range(length)])
+
+
+def hash_string(value, length=16, force_lower=False):
```
Comment (troydai) [inline src/azure-cli-core/azure/cli/core/util.py]: In all of the application of this method below, I didn't see in any case the length is beyond 64, actually none of them uses a name string longer than 16. Therefore, why not just limit the return length to 64. When `length` is greater than 64, which is the sha256 hex digest length, simply throw an exception. By doing so, you reduce the complexity of this method and eliminating the unused behavior.

Comment (tjprescott) [inline src/azure-cli-core/azure/cli/core/util.py]: The original method this was replacing allowed an arbitrary length, so I wanted to ensure that this functionality was preserved in the replacement while still being deterministic.
